packages/grant-explorer/src/context/QFDonationContext.tsx

```diff
@@ -0,0 +1,282 @@
+    voteStatus: context.voteStatus,
+    indexingStatus: context.indexingStatus,
+  };
+};
+
+async function approveTokenForDonation(
+  signerOrProvider: Signer,
+  token: PayoutToken,
+  amount: number,
+  context: QFDonationState
+): Promise<number> {
+  const { setTokenApprovalStatus } = context;
+
+  try {
+    setTokenApprovalStatus(ProgressStatus.IN_PROGRESS);
+
+    const votingStrategy = ethers.constants.AddressZero; // TODO wire in voting contract address
+
+    // TODO: Skip if token is 0x0 as no approval is needed
+
+    // TODO : convert amount to BigNumber for approval
+
+    const { transactionBlockNumber } = await approveTokenOnContract(
+      signerOrProvider,
+      token.address,
+      votingStrategy,
+      amount // TODO: check if this should be BigNum
+    );
+
+    setTokenApprovalStatus(ProgressStatus.IS_SUCCESS);
+
+    return transactionBlockNumber;
+  } catch (error) {
+    datadogLogs.logger.error(
+      `error: approveTokenForDonation - ${error}. Data - ${amount} ${token.name}`
+    );
+    setTokenApprovalStatus(ProgressStatus.IS_ERROR);
+    throw error;
+  }
+}
+
+
+async function vote(
+  signerOrProvider: Signer,
+  roundId: string,
+  token: PayoutToken,
+  donations: FinalBallotDonation[],
+  context: QFDonationState
+): Promise<number> {
+  const { setVoteStatus } = context;
+
+  try {
+    setVoteStatus(ProgressStatus.IN_PROGRESS);
+
+    const encodedVotes = encodeQFVotes(token, donations);
+
+    const { transactionBlockNumber } = await voteOnRoundContract(
+      roundId,
+      signerOrProvider,
+      encodedVotes
+    );
+
+    setVoteStatus(ProgressStatus.IS_SUCCESS);
+    return transactionBlockNumber;
+
+  } catch (error) {
+    datadogLogs.logger.error(
+      `error: approveTokenForDonation - ${error}. Data - ${vote.toString()}`
+    );
+    setVoteStatus(ProgressStatus.IS_ERROR);
+    throw error;
+  }
+
+}
+
+async function waitForSubgraphToUpdate(
+  signerOrProvider: Signer,
+  transactionBlockNumber: number,
+  context: QFDonationState
+) {
+  const { setIndexingStatus } = context;
+
+  try {
+    datadogLogs.logger.error(
+      `waitForSubgraphToUpdate: txnBlockNumber - ${transactionBlockNumber}`
+    );
+
+    setIndexingStatus(ProgressStatus.IN_PROGRESS);
+
+    const chainId = await signerOrProvider.getChainId();
+
+    await waitForSubgraphSyncTo(chainId, transactionBlockNumber);
+
+    setIndexingStatus(ProgressStatus.IS_SUCCESS);
+  } catch (error) {
+    datadogLogs.logger.error(
+      `error: waitForSubgraphToUpdate - ${error}. Data - ${transactionBlockNumber}`
+    );
+    setIndexingStatus(ProgressStatus.IS_ERROR);
+    throw error;
+  }
+}
+
+function encodeQFVotes(donationToken: PayoutToken, donations: FinalBallotDonation[]) : BytesLike[] {
+  let encodedVotes: BytesLike[] = [];
+
+  donations.map(donation => {
+    const vote = [
+      donationToken.address,
+      donation.amount, // TODO: update to BigNum
+      "0x0" // TODO: update to project address
```

Comment:
Right now we store the project registry id but that isn't of much help. We need the payout address